src/libopensc/card-piv.c

```diff
@@ -4493,6 +4494,34 @@ static int piv_get_challenge(sc_card_t *card, u8 *rnd, size_t len)
 
 }
 
+static int piv_authenticate_challenge(sc_card_t *card, const u8 *response_data, size_t response_data_len)
+{
+    u8 sbuf[4096];
+    int r;
+    u8 *p;
+    piv_private_data_t * priv = PIV_DATA(card);
+    
+    LOG_FUNC_CALLED(card->ctx);
+
+    // Build the TLV structure
+    p = sbuf;
+    
+    // Create 7C TLV wrapper
+    r = sc_asn1_put_tag(0x7C, NULL, response_data_len + 2, p, sizeof(sbuf), &p);
+    if (r != SC_SUCCESS)
+        LOG_FUNC_RETURN(card->ctx, r);
+        
+    // Create 82 TLV for the challenge response
+    r = sc_asn1_put_tag(0x82, response_data, response_data_len, p, sizeof(sbuf) - (p - sbuf), &p);
+    if (r != SC_SUCCESS)
+        LOG_FUNC_RETURN(card->ctx, r);
+        
+    // Send GENERAL AUTHENTICATE command
+    r = piv_general_io(card, 0x87, priv->mgmt_key_alg, 0x9B, sbuf, p - sbuf, NULL, 0);
+    
+    LOG_FUNC_RETURN(card->ctx, r);
+}
+
```

Comment:
How is `sc_asn1_put_tag` used in other places to construct nested tags? You have the hard-coded 2 in the implementation, assuming the length is always encoded in one byte. If you start with the innermost tag, you can use the encoded tag length.
```suggestion
static int piv_authenticate_challenge(sc_card_t *card, const u8 *buf, size_t count)
{
	u8 sbuf[4096];
	int r;
	u8 *p;
	piv_private_data_t * priv = PIV_DATA(card);
    
	LOG_FUNC_CALLED(card->ctx);

	// Build the TLV structure
	p = sbuf;
    
	// Create 7C TLV wrapper
	r = sc_asn1_put_tag(0x7C, NULL, count + 2, p, sizeof(sbuf), &p);
	if (r != SC_SUCCESS)
		LOG_FUNC_RETURN(card->ctx, r);
        
	// Create 82 TLV for the challenge response
	r = sc_asn1_put_tag(0x82, buf, count, p, sizeof(sbuf) - (p - sbuf), &p);
	if (r != SC_SUCCESS)
		LOG_FUNC_RETURN(card->ctx, r);
        
	// Send GENERAL AUTHENTICATE command
	r = piv_general_io(card, 0x87, priv->mgmt_key_alg, 0x9B, sbuf, p - sbuf, NULL, 0);
    
	LOG_FUNC_RETURN(card->ctx, r);
}

```

Reply:
good idea